src/hal/src/queue/mod.rs

```diff
@@ -31,32 +34,66 @@ pub enum QueueType {
     Transfer,
 }
 
-impl QueueType {
-    /// Checks for graphics functionality support.
-    /// Supported by general and graphics queues.
-    pub fn supports_graphics(&self) -> bool {
-        *self == QueueType::General || *self == QueueType::Graphics
-    }
+/// General information about a queue family, available upon adapter discovery.
+/// Not useful after an adapter opened and turned into `Gpu`.
+///
+/// *Note*: A backend can expose multiple queue families with the same properties.
+pub trait ProtoQueueFamily: Debug {
+    /// Returns the type of queues.
+    fn queue_type(&self) -> QueueType;
+    /// Returns maximum number of queues created from this family.
+    fn max_queues(&self) -> usize;
+}
+
+/// `RawQueueFamily` denotes a group of command queues provided by the backend
+/// with the same properties/type.
+pub struct RawQueueFamily<B: Backend> {
+    prototype: B::ProtoQueueFamily,
+    queues: Vec<B::CommandQueue>,
+}
 
-    /// Checks for graphics functionality support.
-    /// Supported by general and compute queues.
-    pub fn supports_compute(&self) -> bool {
-        *self == QueueType::General || *self == QueueType::Compute
+//TODO: this is not a very sound structure, unfortunately.
+impl<B: Backend> RawQueueFamily<B> {
+    #[doc(hidden)]
+    pub fn new(prototype: B::ProtoQueueFamily) -> Self {
+        RawQueueFamily {
+            prototype,
+            queues: Vec::new(),
+        }
+    }
+    #[doc(hidden)]
+    pub fn add_queue(&mut self, queue: B::CommandQueue) {
+        assert!(self.queues.len() < self.prototype.max_queues());
+        self.queues.push(queue);
+    }
+    ///
+    pub fn prototype(&self) -> &B::ProtoQueueFamily {
+        &self.prototype
     }
+}
 
-    /// Checks for graphics functionality support.
-    /// Supported by general, graphics and compute queues.
-    pub fn supports_transfer(&self) -> bool { true }
+/// Stronger-typed queue family.
+pub struct QueueFamily<B: Backend, C> {
+    pub(crate) prototype: B::ProtoQueueFamily,
+    /// Command queues created in this family.
+    pub queues: Vec<CommandQueue<B, C>>,
 }
 
-/// `QueueFamily` denotes a group of command queues provided by the backend
-/// with the same properties/type.
-///
-/// *Note*: A backend can expose multiple queue families with the same properties.
-pub trait QueueFamily: 'static {
-    /// Return the number of available queues of this family.
-    // TODO: some backends like d3d12 support infinite software queues (verify)
-    fn num_queues(&self) -> u32;
+impl<B: Backend, C: Capability> QueueFamily<B, C> {
+    /// Create a new strongly typed queue family from a raw one.
+    ///
+    /// *Note*: panics if the family doesn't expose required
```

Comment:
I would prefer the
```
/// # Panics
///
/// Panics if the family doesn't expose required ..
```
doc style as proposed here (https://github.com/rust-lang/rfcs/blob/master/text/1574-more-api-documentation-conventions.md), Fits better here than a note imo.